Parse in status-go and render markdown

cammellos opened the pull request:
Fixes: status-im/trailofbits-audit#47
Fixes: status-im/trailofbits-audit#46
Fixes: status-im/trailofbits-audit#44
Fixes: status-im/security-reports#13
Fixes: status-im/security-reports#5
Fixes: https://github.com/status-im/status-react/issues/8995

This commits re-introduce rendering of markdown text and implent a few
changes:

1) Parsing of the message content is now in status-go, this includes
markdown, line-count, and rtl. Parsing is not nested, as there's some
rendering degradation involved as we nest components, unclear exactly if
it's react-native or clojure, haven't looked too deeply into it.
2) Emojii type messages are now parsed on the receiver side, using the appropriate content-type
3) Fixes a few issues with chat input rendering, currrently we use
`chats/current-chat` subscription which is very heavy and should not be
used unless necessary, and means that
any change to chat will trigger a re-render, which caused re-rendering
of input container on each received message. Also to note that
input-container is fairly heavy to render, and it's rendered twice at
each keypress on input.

Currently group chat messages are handled somewhat differently, and need to be passed to status-go to be parsed before rendering (similarly to outgoing messages). Both flows will be simplified once parsing of group chat updates is in status-go (it is already but seems like it's not working when a message is attached), and sending is fully in status-go.

The inline markdow supported is:

*italic* or _italic_
**bold** or __bold__
`inline code`
http://test.com links
\#status-tag

The block markdown supported is:

\# Headers
```
code blocks
```
> Quotereply

The styling is very basic at the moment, but can be improved.
Adding other markdown (photo,mentions) is straightforward and should
come at little performance cost (unless the component to render is
heavy, i.e a photo for example).

There are some behavioral changes with this commit:

1) Links are only parsed if starting with http:// or https://, meaning that
blah.com won't be parsed, nor www.test.com. This behavior is consistent
with discord for example and allows faster parsing at little expense to
ser experience imo. Fixes a few security issues as well.

2) Content is not anymore capped (regression), that's due to the fact that
before we only rendered text and react-native allowed us easily to limit
the number of lines, but adding markdown support means that this
strategy is not viable anymore. Performance of rendering don't see to be
very much impacted by this, I would re-introduce it if necessary, but
I'd rather do that in a separate PR.

cc @errorists @rachelhamlin 

status: ready
![1-a](https://user-images.githubusercontent.com/1017008/68770830-2d328c00-0627-11ea-9ea9-d1467ea636a7.png)
![2-a](https://user-images.githubusercontent.com/1017008/68770832-2dcb2280-0627-11ea-96f4-76ac7ed3d39f.png)

cammellos committed:
Render markdown

Fixes: https://github.com/status-im/trailofbits-audit/issues/47
Fixes: https://github.com/status-im/trailofbits-audit/issues/46
Fixes: https://github.com/status-im/trailofbits-audit/issues/44
Fixes: https://github.com/status-im/security-reports/issues/13
Fixes: https://github.com/status-im/security-reports/issues/5
Fixes: https://github.com/status-im/status-react/issues/8995

This commits re-introduce rendering of markdown text and implent a few
changes:

1) Parsing of the message content is now in status-go, this includes
markdown, line-count, and rtl. Parsing is not nested, as there's some
rendering degradation involved as we nest components, unclear exactly if
it's react-native or clojure, haven't looked too deeply into it.
2) Emojii type messages are not parsed on the sending side, not the
receiving one, using the appropriate content-type
3) Fixes a few issues with chat input rendering, currrently we use
`chats/current-chat` subscription which is very heavy and should not be
used unless necessary, and means that
any change to chat will trigger a re-render, which caused re-rendering
of input container on each received message. Also to note that
input-container is fairly heavy to render, and it's rendered twice at
each keypress on input.

The inline markdow supported is:

*italic* or _italic_
**bold** or __bold__
`inline code`
http://test.com links
\#status-tag

The block markdown supported is:

\# Headers
```
code blocks
```
> Quotereply

The styling is very basic at the moment, but can be improved.
Adding other markdown (photo,mentions) is straightforward and should
come at little performance cost (unless the component to render is
heavy, i.e a photo for example).

There are some behavioral changes with this commit:

1) Links are only parsed if starting with http:// or https://, meaning that
blah.com won't be parsed, nor www.test.com. This behavior is consistent
with discord for example and allows faster parsing at little expense to
ser experience imo. Fixes a few security issues as well.

2) Content is not anymore capped (regression), that's due to the fact that
before we only rendered text and react-native allowed us easily to limit
the number of lines, but adding markdown support means that this
strategy is not viable anymore. Performance of rendering don't see to be
very much impacted by this, I would re-introduce it if necessary, but
I'd rather do that in a separate PR.

Signed-off-by: Andrea Maria Piana <andrea.maria.piana@gmail.com>

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 6cb88db6 | [#2](https://ci.status.im/job/status-react/job/prs/job/ios/job/PR-9409/2/) | 2019-11-07 14:00:34 | ~12 min | `ios` | [:package:`ipa`](https://i.diawi.com/o5RYPr) [:calling:](https://chart.apis.google.com/chart?cht=qr&chs=400x400&chld=L%7C%0A1&chl=https%3A%2F%2Fi.diawi.com%2Fo5RYPr)|

cammellos commented:
@Serhy I Have re-introduced markdown (`code`, **bold**, *italic*, http://link.com, and #tag)  and fixed a few security issues in this PR:
https://github.com/status-im/trailofbits-audit/issues/46 
https://github.com/status-im/security-reports/issues/13
and https://github.com/status-im/trailofbits-audit/issues/44

it's still a bit rough and there are few bugs here and there which I will address,
would you mind if you have time comparing rendering performance with develop to check that performance rendering messages is comparable (it seems like it is to me, but worth double checking). Some performance degradation is probably inevitable as current develop does not handle italic/bold/code but only links and #tag , but hopefully it's roughly the same.

Thanks!

cammellos commented:
@Serhy 
I have added support for some more markdown, could you check perf of the last build?

i have added support for:
```
# heading

> quote replies

\``` (remove the slash)
code blocks
\```

in addition to
**bold**
*italic*
http;//link.com
#status-tag
```

Thanks!

Serhy commented:
Tested performance behavior of message rendering of nested a) links and b) tags in this PR and current nightly develop build.
When user of Galaxy Note 4 navigates to public chat having 5 messages (2500 chars each with 250 items) time between tap and when messages appear in chat view takes:

Links (`#links3` public channel)

| build/attempt  |  PR build (StatusIm-191112-103835-89432a-pr-universal.apk| 12th Nov nightly build |
| ------------- | ------------- | ------------- |
| Navigate with unread msg  | 6.6s | 6.6s  |
| Second navigate  | 6.4s  | 6.2s |
| Third navigate  |  6.3s | 6.3s  |
| **Average** | **6.43s**  | **6.36s** |


Tags (`#tags` public channel)

| build/attempt  |  PR build (StatusIm-191112-103835-89432a-pr-universal.apk| 12th Nov nightly build |
| ------------- | ------------- | ------------- |
| Navigate with unread msg  | 9.15s | 8.75s  |
| Second navigate  |  8.6s | 8.5s |
| Third navigate  | 8.8s  | 8.33s  |
| **Average** |  **8.85s**  | **8.52s** |

Almost no changes (which I understand as good result here). 
It makes sense to note that this morning I was checking also previous build (8th in Jenkins which is https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-191111-140910-277801-pr-universal.apk) and it's a bit faster for results. For example the same scenario with public chat having `#tags` takes ~6.8s (about 20% faster than develop)

> @Serhy
> I have added support for some more markdown, could you check perf of the last build?

Will check quotes, code, bold/italic formatting with next  but no way to compare with current develop since we don't have such markdown for develop.

Serhy commented:
Below are the figures for markdown with 5 text messages, each message has 2500 chars and 250 items to parse. With the Galaxy Note 4 and https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-191112-130525-fd4943-pr-universal.apk build takes:

| Navigate to chat attempt | Bold | Italic | Code quote | Quote | Header |
| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
| First (unread messages) | 3.5s  | 3.9s | 5.6s | 1s | 4.3s | 
| Second | 3.5s | 3.9s | 6s | 0.9s | 4.8s | 
| Third | 3.6s  | 3.8s | 6.15s | 1.03  | 5.2s |
| Average | **3.53s**  | **3.86s** | **5.91** | **0.97s** | **4.76s** |

Can't compare this figures with current develop because we don't have bold/italic/code markdown in develop and even more quotes and headers. 

The last figures for performance measurements i found in https://github.com/status-im/status-react/issues/8263 (but it's before this issue fixed as closed). 

| Bold | Italic | Backquotes | 
| ------------- | ------------- | ------------- |
| 16s | 16s | 26s |

cammellos commented:
Thanks @Serhy! Would you say this are acceptable performance for develop? If so I can go ahead and polish the PR so that it gets merged.
Thanks!

Serhy commented:
Yes, @cammellos , absolutely. There is ~2% perf degradation between PR and Develop build, though it's not a big deal taking into account margins of error.

errorists commented:
cool! 
I think that `#headers` are a step too far, we're at a limit here of what looks 'normal' in the context of message bubble containers. headers looks out of place, now if we _really_ want to go there I'd suggest a bubble-less message appearance. Same with things like lists.
for now, I'd ask if we could discuss removing headers? @rachelhamlin 

obviously code blocks needs styling, so do inline quotes
what I had in mind was more like so
<img width="495" alt="Screenshot 2019-11-13 at 12 24 57" src="https://user-images.githubusercontent.com/4932774/68759763-9f4ba680-0610-11ea-8630-33506ee29d74.png">

cammellos commented:
> I'd ask if we could discuss removing headers? @rachelhamlin

That's fine with me, unless we really want them

I can style the codeblock (minus the syntax highlighting, that's another can of worms :) ) , inline quotes as well, do you have styling though for the outgoing quote? (i.e the messages i send)

errorists commented:
thanks @cammellos yea syntax highlighting was a long shot, wanted to make you engineering guys happy :D 

the code block:
background: `#2E386B`
text color: white
border: none
corner radius: 4px
font-family: let's do `monospace` for now, in future we'd like to look into deploying Roboto Mono specifically across all platforms
font-size / font-weight / line-height should all be inherited from regular text

<img width="277" alt="inline-incoming" src="https://user-images.githubusercontent.com/4932774/68761083-9c9e8080-0613-11ea-9d0c-1b565ec01804.png">

Inline quotes, incoming:
2px left border, border color: `Grey` from our palette (same as timestamps)
space between the border and text: 6px
font-size: 14
line-height: 19
text-color: `rgba(0,0,0, 0.5)` - so that's black with 50% opacity
2px of additional vertical margin between the quote and any regular text
font-family and font-weight are inherited from regular text

<img width="277" alt="inline-outgoing" src="https://user-images.githubusercontent.com/4932774/68761060-8ee8fb00-0613-11ea-8779-a60fe869951a.png">

Outgoing inline quote
border-color: `40% white` from our existing colour palette, that's `rgba(255,255,255, 0.4)`
text-colour: `70% white` from our colour palette, that's `rgba(255,255,255, 0.7)`

cammellos commented:
@errorists updated

corpetty commented on src/status_im/chat/models/input.cljs:
```diff
@@ -125,11 +126,15 @@
                 {:db (assoc-in db [:chats current-chat-id :metadata :responding-to-message] nil)}
                 (chat.message/send-message {:chat-id      current-chat-id
-                                            :content-type constants/content-type-text
+                                            :content-type (if emoji?
```
How does this work in the context of mixed text/emoji?

cammellos replied:
oh yeah good point, it gets rendered as massive text, probably best to validate that message, but will be in status-go, I am moving validation there, so I might do it in the scope of a separate PR

corpetty commented on test/cljs/status_im/test/utils/security.cljs:
```diff
@@ -16,6 +16,10 @@
   (testing "a javascript link"
     (is (not (security/safe-link? "javascript://anything"))))
+  (testing "a javascript link mixed cases"
+    (is (not (security/safe-link? "JaVasCrIpt://anything"))))
```
if we're no longer rendering these as clickable, are they necessary?

corpetty commented on src/status_im/utils/security.cljs:
```diff
@@ -24,7 +24,7 @@
 
 ;; Links starting with javascript:// should not be handled at all
-(def javascript-link-regex #"javascript://.*")
+(def javascript-link-regex #"(?i)javascript://.*")
```
if we aren't rendering anything but `http` and `https` as clickable, are these needed?

cammellos replied:
technically no, but given that status-go is a separate component, if there's any regression there, then status-react would blindly follow those links. This code is only run once the user clicks on a link, so performance is not impacted, therefore I would probably keep it, just to be on the safe side. what do you think?

corpetty replied:
I don't mind having multiple security checks like this.. i'd rather over validate than under.

corpetty reviewed (commented):
A few comments on questions.

cammellos commented:
@churik @Serhy I have updated the link test as we don't support status.im type links anymore, only http://status.im or https://status.im, the other failures don't seem to be related to this PR, could you please have a quick look?
Thanks!

rasom reviewed (approved):
Can we make 80+ lines shorter? :)
